skimage/feature/register_translation.py

```diff
@@ -59,21 +59,26 @@ def _upsampled_dft(data, upsampled_region_size,
             raise ValueError("number of axis offsets must be equal to input "
                              "data's number of dimensions.")
 
-    col_kernel = np.exp(
-        (-1j * 2 * np.pi / (data.shape[1] * upsample_factor)) *
-        (np.fft.ifftshift(np.arange(data.shape[1]))[:, None] -
-         np.floor(data.shape[1] / 2)).dot(
-             np.arange(upsampled_region_size[1])[None, :] - axis_offsets[1])
-    )
-    row_kernel = np.exp(
-        (-1j * 2 * np.pi / (data.shape[0] * upsample_factor)) *
-        (np.arange(upsampled_region_size[0])[:, None] - axis_offsets[0]).dot(
-            np.fft.ifftshift(np.arange(data.shape[0]))[None, :] -
-            np.floor(data.shape[0] / 2))
-    )
+    im2pi = 1j * 2 * np.pi
 
-    return row_kernel.dot(data).dot(col_kernel)
+    dim_kernels = []
+    for (n_items, ups_size, ax_offset) in zip(data.shape,
+                                              upsampled_region_size,
+                                              axis_offsets):
+        dim_kernels.append(
+            np.exp((-im2pi / (n_items * upsample_factor)) *
+                   (np.arange(upsampled_region_size[0])[:, None] - ax_offset) @
+                   (np.fft.ifftshift(np.arange(n_items))[None,
+                    :] - n_items // 2))
 
+        )
+
+    if data.ndim == 2:
+        einsum_string = 'ij, ki, lj'
+    elif data.ndim == 3:
+        einsum_string = 'ijk, li, mj, nk'
```

Comment:
Can you add the output strings to this? Explicit is better than implicit. =) Also, could you time the einsum version of the kernel convolution compared to the original `np.dot` version? I suspect `.dot` will be much faster, in which case it would be worth keeping it for the 2D case, and only using einsum for 3D.